source/Mlos.Python/mlos/Optimizers/OptimizerBase.py

```diff
@@ -91,10 +94,22 @@ def optimum(self, optimum_definition: OptimumDefinition = OptimumDefinition.BEST
         if not len(features_df.index):
             raise ValueError("Can't compute optimum before registering any observations.")
 
+        if context is not None and optimum_definition != OptimumDefinition.BEST_PREDICTED_WITHIN_CONTEXT:
+                raise ValueError(f"{optimum_definition} not supported if context is provided.")
+
         if optimum_definition == OptimumDefinition.BEST_OBSERVATION:
             return self._best_observation_optimum(features_df=features_df, objectives_df=objectives_df)
+        elif optimum_definition == OptimumDefinition.BEST_PREDICTED_WITHIN_CONTEXT:
+            if context is None:
+                raise ValueError(f"{optimum_definition} requires context to be not None.")
+            return self._optimum_within_context(context=context)
         return self._prediction_based_optimum(features_df=features_df, optimum_definition=optimum_definition, alpha=alpha)
 
+    @trace()
+    def _optimum_within_context(self, context: pd.DataFrame):
+        return self.experiment_designer.numeric_optimizer.maximize(

```

Comment:
Is this returning speculative configurations? In other words, can it return a config we never tried?

Reply:
Yes. I'm not entirely sure what else we should do. If the context is continuous, then we likely will not have any configurations for that context, right?
or do you mean configurations independent of context, like configurations we tried with other context? That might be another plausible method. It would still be speculative in the sense that the combination of configuration and context would be new.

Reply:
I think there are production scenarios where one may prefer to only use points that have actually been sampled.  Basically, it should be a configurable constraint on what to return.

Definitely like your other comment about returning speculative configs for new contexts though (e.g. a machine SKU you haven't tested yet, but is similar to an existing one).

Reply:
So we are thinking of the context as discrete? Then I might reconsider some of the assumptions.
If we look at featured derived from a workload as the context, it's unlikely we'll observe all of them, right?